File: typescript/src/age.ts
import { ymd } from "./dates.js";

export interface Tenure {
  years: number;
  months: number;
}

function totalMonths(from: string | Date, to: string | Date): number {
  const a = ymd(from);
  const b = ymd(to);
  let months = (b.y - a.y) * 12 + (b.m - a.m);
  if (b.d < a.d) months -= 1;
  return months;
}

/** from → to の満年・月（勤続年数など）。Dart/Java 版と同一。 */
export function tenure(from: string | Date, to: string | Date): Tenure {
  const t = totalMonths(from, to);
  return { years: Math.trunc(t / 12), months: t % 12 };
}

/** asOf 時点の満年齢。 */
export function ageAt(birth: string | Date, asOf: string | Date): number {
  return tenure(birth, asOf).years;
}

File: java/src/main/java/io/hatake/core/Age.java
package io.hatake.core;

import java.time.LocalDate;

/** 年齢・勤続年数。Dart / TypeScript 版と同一。 */
public final class Age {
    private Age() {
    }

    public record Tenure(int years, int months) {
    }

    public static Tenure tenure(String from, String to) {
        LocalDate a = Dates.parse(from);
        LocalDate b = Dates.parse(to);
        int months = (b.getYear() - a.getYear()) * 12 + (b.getMonthValue() - a.getMonthValue());
        if (b.getDayOfMonth() < a.getDayOfMonth()) {
            months -= 1;
        }
        return new Tenure(months / 12, months % 12);
    }

    public static int ageAt(String birth, String asOf) {
        return tenure(birth, asOf).years();
    }
}
